[quant] avoid resize calls in observer/fake_quant

supriyar opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#60386 [quant] avoid resize calls in observer/fake_quant**

Summary:
During QAT we sometimes encounter errors with scripted models
`RuntimeError: cannot resize variables that require grad`

For per-tensor cases we don't need to resize some buffers so this PR removes the extra resize ops where applicable
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29271905](https://our.internmc.facebook.com/intern/diff/D29271905)

supriyar committed:
[quant] avoid resize calls in observer/fake_quant

Summary:
During QAT we sometimes encounter errors with scripted models
`RuntimeError: cannot resize variables that require grad`

For per-tensor cases we don't need to resize some buffers so this PR removes the extra resize ops where applicable
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29271905).

vkuzo commented on torch/quantization/observer.py:
```diff
@@ -533,8 +533,6 @@ def forward(self, x_orig):
             min_val_cur, max_val_cur = torch._aminmax(x)
             min_val = min_val + self.averaging_constant * (min_val_cur - min_val)
             max_val = max_val + self.averaging_constant * (max_val_cur - max_val)
-        self.min_val.resize_(min_val.shape)
```
can we keep this and gate behind an if statement, like we did for `FakeQuantize?

supriyar replied:
For this observer we will always have the min/max val have a shape of 1, I believe. We have a separate one - `MovingAveragePerChannelMinMaxObserver` which has different shape.

vkuzo reviewed (approved):
lg if CI passes

supriyar committed:
Update on "[quant] avoid resize calls in observer/fake_quant"


Summary:
During QAT we sometimes encounter errors with scripted models
`RuntimeError: cannot resize variables that require grad`

For per-tensor cases we don't need to resize some buffers so this PR removes the extra resize ops where applicable
Test Plan:

Reviewers:

Subscribers:

Tasks:

Tags:

Differential Revision: [D29271905](https://our.internmc.facebook.com/intern/diff/D29271905)

[ghstack-poisoned]

supriyar commented:
@supriyar has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D29271905).
